### PR TITLE
Add hazelcast-client extension

### PR DESCRIPTION
**Pull request by pivovarit**
Hi!

I prepared the initial version of the `hazelcast-client` extension. 

Design-wise, we need to resolve and parse configuration files at runtime in order to maintain environment independence. We could push most of it to build-time gaining a few milliseconds, but we'd need to rebuild the native image for each environment.

Please review and advise.

----
Related: https://github.com/quarkusio/quarkus/pull/6792

### CODE REVIEW

**Comment by pivovarit** [comment]
Should I add the `hazelcast-client` module to one of the pipelines?

**Comment by geoand** [comment]
Thanks a lot for this.

@pivovarit yes please add it. Otherwise the native tests won't be run.

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (geoand) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Please remove the author tags, we have decided to not add them to new code.

**Comment by pivovarit** [comment]
> Thanks a lot for this.
> 
> @pivovarit yes please add it. Otherwise the native tests won't be run.

You are welcome!

I believe it's already running. If I just `mvn clean install -Dquarkus.native.container-build=true ` everything looks good. 

However, I still haven't deciphered your pipelines so I might be missing a place or two, could you point me to the right place?

**Comment by geoand** [comment]
> 
> However, I still haven't deciphered your pipelines so I might be missing a place or two, could you point me to the right place?

Here is the place where we define the various native tests: https://github.com/quarkusio/quarkus/blob/master/ci-templates/stages.yml#L190

File: extensions/hazelcast-client/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gastaldi) [inline extensions/hazelcast-client/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: You need to add a `hazelcast-client.adoc` in https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc to avoid the broken link

File: integration-tests/hazelcast-client/src/main/java/io/quarkus/it/hazelcast/client/CustomNodeFilter.java
Comment (gastaldi) [inline integration-tests/hazelcast-client/src/main/java/io/quarkus/it/hazelcast/client/CustomNodeFilter.java]: Remove any call to System.out in tests

File: extensions/hazelcast-client/runtime/pom.xml
Comment (gastaldi) [inline extensions/hazelcast-client/runtime/pom.xml]: Add these versions to the bom/runtime/pom.xml BOM

Comment (gastaldi) [reply extensions/hazelcast-client/runtime/pom.xml]: Also out of curiosity, do you need the `-all` dependency or can you use separate artifacts? It looks like it's 12Mb and I am not sure if all these classes are required, as they would impact the size of the resulting application

Comment (pivovarit) [reply extensions/hazelcast-client/runtime/pom.xml]: In a new 4.0 release, we'll be merging some artifacts, but I believe for 3.0 we should be able to strip it down a bit

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (gastaldi) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: It looks like they are already handled in the `jaxb` extension? Try depending your extension on it and see if that's sufficient.

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (gastaldi) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Move this to the `FeatureBuildItem`  class for now

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (gastaldi) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: As a good practice, avoid storing the BuildProducers as attributes and pass them as parameters in the methods you need them. This helps extension maintainers to understand what's necessary in each method

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (gastaldi) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Not a big deal, but you can just return the instantiated BuildItems in these methods :)

File: integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java
Comment (gastaldi) [inline integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: Maybe call member1.shutdown() just in case?

**Comment by geoand** [comment]
`master` (hopefully) contains a fix for the CI issue so rebasing onto it is advised

**Comment by pivovarit** [comment]
@geoand sure thing, the work here will slow down a bit since I'm reworking a few things - we made a decision to target the next version of `hazelcast` which requires a couple more native mode fixes

**Comment by geoand** [comment]
Sure thing, thanks for the update!

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (pivovarit) [reply extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Unfortunately, it doesn't really help - I will investigate and ideally fix the issues in `jaxb` so that we can reuse it

File: extensions/hazelcast-client/README.MD
Comment (gsmet) [inline extensions/hazelcast-client/README.MD]: This is outdated as we now only support 19.3.1 in master and that's my problem with having a README here: it will quickly be outdated. All this should be in the guide we publish to the website instead IMHO. We have attributes there and they are probably updated for each release.

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (gsmet) [reply extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: +1

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (gsmet) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: You can drop that one, it's not needed anymore in master as 19.3 has JNI always enabled.

File: extensions/hazelcast-client/runtime/pom.xml
Comment (gsmet) [inline extensions/hazelcast-client/runtime/pom.xml]: Same here it should be in the BOM. But it doesn't come with Hazelcast? I'm surprised you even need to add it.

I'm also totally unsure it's useful as we don't have JSR 107 support in Quarkus.

File: extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastClientConfig.java
Comment (gsmet) [inline extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastClientConfig.java]: Can't they be a list of `InetSocketAddress`?

It would be better to not have it optional and have a default value instead (if possible). That way, it's automatically documented.

File: extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastClientConfig.java
Comment (gsmet) [inline extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastClientConfig.java]: Better use `OptionalInt` for all of these.

File: extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastConfigurationParser.java
Comment (gsmet) [inline extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastConfigurationParser.java]: We avoid lambdas in runtime code. They have a cost.

Comment (gsmet) [inline extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastConfigurationParser.java]: Your config should be a List and the split on `,` will be done automatically by the config framework.

File: extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastConfigurationParser.java
Comment (gsmet) [inline extensions/hazelcast-client/runtime/src/main/java/io/quarkus/hazelcast/client/runtime/HazelcastConfigurationParser.java]: Same here and for all the other elements that are split.

File: extensions/hazelcast-client/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gsmet) [inline extensions/hazelcast-client/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: Let's make it `preview` first.

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks for this work. I added a couple of comments.

**Comment by pivovarit** [comment]
@gsmet thanks for all the comments, keep in mind it's still work in progress - I will call all of you once all the changes are done - I will wait a week or two for the Hazelcast 4.0.0

File: extensions/hazelcast-client/README.MD
Comment (pivovarit) [reply extensions/hazelcast-client/README.MD]: I agree, I know already that there's a separate place for guides and I'm planning to move it all there!

File: extensions/hazelcast-client/runtime/pom.xml
Comment (pivovarit) [reply extensions/hazelcast-client/runtime/pom.xml]: Hazelcast 3.0.0 has `cache-api` in as _provided_ dependency - most likely we should be able to get rid of it in the next release where we're reshaping our artifacts

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (pivovarit) [reply extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Fix'd: https://github.com/quarkusio/quarkus/pull/6863/files

**Comment by pivovarit** [comment]
[Hazelcast 4.0 is already in Maven Central](https://search.maven.org/artifact/com.hazelcast/hazelcast/4.0/jar) so I will update the PR and it should be ready today/tomorrow for the ultimate review :)

File: integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java
Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: All the client instances get registered internally so I'd say this is the "just in case" option

**Comment by pivovarit** [comment]
It's almost ready. However, I need to consult one thing.

At the moment, the bean produced by the producer is initialized lazily (I believe it's CDI's default).
This results in our client establishing a connection on the first call touching the client bean.

Not sure if this is desired or not. Should we leave the choice to users and leave it as it or force eager init?

**Comment by geoand** [comment]
> It's almost ready. However, I need to consult a few things.
> 
> At the moment, the bean produced by the producer is initialized lazily (I believe it's CDI's default).
> This results in our client establishing a connection on the first call touching the client bean.
> 
> Not sure if this is desired or not. Should we leave the choice to users and leave it as it or force eager init?
> 

@karesti how is Infinispan handled? I think we should follow the same model here.

> Another thing, there's no separation between `hazelcast` and `hazelcast-client` anymore, it's just `hazelcast` since 4.0. Is it ok if we provide both embedded and client functionality through a single `quarkus-hazelcast` extension?

I am pretty sure that for Infinispan there are separate extensions and I think we should try to be consistent. Again @karesti can provide more input here :)

File: extensions/hazelcast-client/runtime/pom.xml
Comment (pivovarit) [reply extensions/hazelcast-client/runtime/pom.xml]: That's no longer the case in 4.0, we are on the bare minimum.

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (machi1990) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Maybe also watch these files for modification with `HotDeploymentWatchedFileBuildItem`?

Comment (pivovarit) [reply extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Good idea, thanks!

File: extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java
Comment (machi1990) [inline extensions/hazelcast-client/deployment/src/main/java/io/quarkus/hazelcast/client/deployment/HazelcastClientProcessor.java]: Current master is on 19.3.1, I guess we can remove this.

File: extensions/hazelcast-client/deployment/pom.xml
Comment (gastaldi) [inline extensions/hazelcast-client/deployment/pom.xml]: You can remove this if you have added `quarkus-hazelcast-client` to the `bom/runtime/pom.xml`

File: extensions/hazelcast-client/runtime/pom.xml
Comment (gastaldi) [reply extensions/hazelcast-client/runtime/pom.xml]: Is this dependency still needed now that Hazelcast Client 4.0 is used?

**Review by gastaldi — CHANGES_REQUESTED** [review]
Looks good. Added some minor comments that would be nice to be addressed before merging this

File: extensions/hazelcast-client/runtime/pom.xml
Comment (pivovarit) [reply extensions/hazelcast-client/runtime/pom.xml]: Unfortunately, yes. The dependency is not used for anything other than making GraalVM happy, and manual removal of it from core classes via substitutions would last weeks and be super fragile.

The real hope to get rid of is if this gets implemented: https://github.com/oracle/graal/issues/1664 and then similar feature could be exposed through Quarkus build item

**Review by gastaldi — APPROVED** [review]
LGTM

File: extensions/hazelcast-client/runtime/pom.xml
Comment (geoand) [reply extensions/hazelcast-client/runtime/pom.xml]: Are you sure that stripping out the classes via substitutions would be that difficult? We really try to minimize runtime dependencies as much as possible.

Comment (pivovarit) [reply extensions/hazelcast-client/runtime/pom.xml]: To give you a clue what the volume of changes is: https://github.com/hazelcast/hazelcast/search?q=javax.cache&unscoped_q=javax.cache

We're totally coupled to jcache

Comment (geoand) [reply extensions/hazelcast-client/runtime/pom.xml]: I see, thanks

**Comment by wburns** [comment]
> > It's almost ready. However, I need to consult a few things.
> > At the moment, the bean produced by the producer is initialized lazily (I believe it's CDI's default).
> > This results in our client establishing a connection on the first call touching the client bean.
> > Not sure if this is desired or not. Should we leave the choice to users and leave it as it or force eager init?
> 
> @karesti how is Infinispan handled? I think we should follow the same model here.

This is the same for Infinispan Client. The connection isn't initialized until the bean is referenced.

> 
> > Another thing, there's no separation between `hazelcast` and `hazelcast-client` anymore, it's just `hazelcast` since 4.0. Is it ok if we provide both embedded and client functionality through a single `quarkus-hazelcast` extension?
> 
> I am pretty sure that for Infinispan there are separate extensions and I think we should try to be consistent. Again @karesti can provide more input here :)

I would think you would want to have them separate. In the Infinispan case, the client is a lot leaner than the embedded extension so we don't want them together.

**Comment by pivovarit** [comment]
@wburns thanks, I took the same approach

File: extensions/hazelcast-client/runtime/pom.xml
Comment (pivovarit) [reply extensions/hazelcast-client/runtime/pom.xml]: What we could do instead is to make the dependency optional and include a note in the README that it's required for native... let me know if this is ok

Comment (geoand) [reply extensions/hazelcast-client/runtime/pom.xml]: I think it's probably better as is. We really try to make the switch from jvm mode to native just a matter of adding `-Dnative`.

File: integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java
Comment (geoand) [inline integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: :+1:

Comment (geoand) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: Perhaps this could also be mentioned in the doc? Otherwise I think it will be unlikely most users will discover it.

Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: That's just a part of testing infrastructure under `src/test`, I believe that should not be exposed. If it is somehow, it's a bug, I believe

Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: Used only here: https://github.com/quarkusio/quarkus/blob/6b7fbe650cd3fee005ad6b94485f21491ac1053b/integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastClientFunctionalityTest.java#L12

Comment (geoand) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: Yeah, but aren't users supposed to be able to use this resource? I mean the other resources we have users can use and are encouraged to use actually

Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: I did not really realize it's going to be public, but well, no problem :) I will double-check and add a note in the docs

Comment (geoand) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: See for example: `DerbyDatabaseTestResource` or `H2DatabaseTestResource`. I mean it's up to you if you want to expose it or not.
But wouldn't it be useful for users to have?

Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: Actually, it's quite useful :) yeah, would love to move forward with that

Comment (geoand) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: Excellent!

Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: I exposed the functionality via `quarkus-test-hazelcast` :)

Comment (geoand) [reply integration-tests/hazelcast-client/src/test/java/io/quarkus/it/hazelcast/client/HazelcastServerTestResource.java]: With accompanying documentation, :+1:

**Comment by geoand** [comment]
@pivovarit Thank you a lot for this work! 

I'm sorry to let you know that because we are in a real sprint to get 1.3 out of the door, I am unlikely to be able to take another good look at this before then. 
I'll circle back as soon as possible.

Thanks for your patience!

**Comment by pivovarit** [comment]
Just a random thought here. After this gets in, I'd love to integrate it with Hibernate 2LC alongside Caffeine and Infinispan

**Comment by pivovarit** [comment]
Can I help you somehow with pushing this forward? :)

btw, I just updated the newly introduced GitHub Actions pipelines.

**Comment by emmanuelbernard** [comment]
Hey @pivovarit 

Thanks for the PR!

tl;dr; We want the extension to be in Quarkus but that does not mean it has to go in the quarkus repo (we call it incorrectly quarkus-core).

Longer version
So, we definitely want the Hazelcast client extension to be present in the ecosystem (code.quarkus.io being the main window of it). We tried the "monorepo" approach but it is straining our systems too much (humans to keep track of everything, machines e.g. CI - we are close to the GH action limits already, etc). So what we are doing now is move new extensions to their own GH repos and integrate them in quarkus-platform (which is the BOM + metadata we use to create a quarkus release) https://github.com/quarkusio/quarkus-platform/
Right now the Debezium outbox pattern and the JMS for QPid extensions are being hosted that way. This is still a bit of work in progress but that's where we are aiming and will help us to scale.

For a user, there is no difference between extensions hosted on quarkusio/quarkus or elsewhere. In particular, the doc and code.quarkus.io presence is the same. For an extension developer, the difference is that
- the release cycle is more in your control
- issue management is in your control too
- it's not part of the initial CI build upon each change but now that we have moved to GH actions, we believe it will be easy to chain GH Action builds and thus trigger a platform extension build after core has changed

I am sorry it's taking so long and for this extra step. If you all are ok with this plan, my hope is for the hazelcast extension to be amongst the early adopters of this model (with QPid and Debezium).

BTW another point, you plan to keep maintaining it, right? If an extension is part of the platform, sometimes we change extension contracts to keep the core evolving and we would need your help in keeping it up to date to ensure the platform does not lose features.

**Comment by geoand** [comment]
@pivovarit Thank you very much for your hard work and understanding on this. I'll also like to express my sympathies for you getting caught up in the middle of our repo reorganization.

Following up on what @emmanuelbernard mentions, there are a few simple concrete steps that you can follow to make your extension part of the Quarkus platform.

* Create your own repo to host the extension
* Open a PR against the Quarkus Platform repo: https://github.com/quarkusio/quarkus-platform

The good thing is that there are already two PRs in flight doing the exact same thing: https://github.com/quarkusio/quarkus-platform/pull/28, https://github.com/quarkusio/quarkus-platform/pull/25 so you should be have some good examples of what needs to be done.
And of course any help you need, don't hesitate to ping us

**Comment by pivovarit** [comment]
Thank you for the update - I feel the monorepo pain and will gladly help you with migrating to the new setup. I will move the content to a separate repository and integrate it with the platform soon.

**Comment by pivovarit** [comment]
Superseded with: https://github.com/hazelcast/quarkus-hazelcast-client/pull/1

**Comment by pivovarit** [comment]
@geoand I have one question. Where should I add the documentation so that it's visible on the Quarkus site?

**Comment by geoand** [comment]
@gsmet would be the most qualified to answer this question :)

**Comment by gastaldi** [comment]
@pivovarit docs go in https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc for now. Since it's a community extension, It should preferably just contain a reference to the docs maintained in your repository (See https://github.com/quarkusio/quarkus/blob/master/docs/src/main/asciidoc/camel.adoc)

**Pull request closed without merging**
